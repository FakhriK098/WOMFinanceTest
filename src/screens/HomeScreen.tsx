import React from 'react';
import { StyleSheet, View, useColorScheme } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import Text from '../components/atoms/Text';
import Button from '../components/atoms/Button';
import { logout } from '../store/slices/authSlice';
import { RootState, AppDispatch } from '../store';
import { colors } from '../theme/colors';
import { spacing } from '../theme/spacing';
import { typography } from '../theme/typography';
import Spacer from '../components/atoms/Spacer';

const HomeScreen = () => {
  const dispatch = useDispatch<AppDispatch>();
  const email = useSelector((s: RootState) => s.auth.email);
  const isDark = useColorScheme() === 'dark';
  return (
    <View
      style={[
        styles.wrapper,
        {
          backgroundColor: isDark
            ? colors.background.dark
            : colors.background.light,
        },
      ]}
    >
      <Text
        weight="bold"
        size={typography.xl}
        color={isDark ? colors.text.dark : colors.text.light}
      >
        Pokedex
      </Text>
      <Spacer size={spacing.lg} />
      <Text
        size={typography.sm}
        color={isDark ? colors.text.dark : colors.text.light}
      >
        Masuk sebagai {email}
      </Text>
      <Spacer size={spacing.lg} />
      <Button title="Keluar" onPress={() => dispatch(logout())} />
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    padding: spacing.lg,
    justifyContent: 'center',
  },
});

export default HomeScreen;
