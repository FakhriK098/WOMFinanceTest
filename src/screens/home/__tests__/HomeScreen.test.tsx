import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { useSelector, useDispatch } from 'react-redux';
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
import * as RN from 'react-native';

// Mock navigation
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: jest.fn(), setOptions: jest.fn() }),
}));
jest.mock('@store', () => ({ store: {} }));
jest.mock('@store/slices/authSlice', () => ({ logout: () => ({ type: 'logout' }) }));
jest.mock('@store/slices/pokemonSlice', () => ({ fetchPokemons: () => ({ type: 'fetchPokemons' }) }));

import HomeScreen from '../HomeScreen';

describe('HomeScreen', () => {
  const useSelectorSpy = useSelector as unknown as jest.Mock;
  const useDispatchSpy = useDispatch as unknown as jest.Mock;

  beforeEach(() => {
    jest.spyOn(RN, 'useColorScheme').mockReturnValue('light');
    useDispatchSpy.mockReturnValue(jest.fn());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('shows initial loading state (snapshot)', () => {
    useSelectorSpy.mockImplementation((sel: any) =>
      sel({
        auth: { email: 'user@example.com' },
        pokemon: {
          items: [],
          loading: true,
          refreshing: false,
          error: null,
          nextUrl: null,
        },
      }),
    );
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<HomeScreen />);
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('renders error state with retry (snapshot)', () => {
    useSelectorSpy.mockImplementation((sel: any) =>
      sel({
        auth: { email: 'user@example.com' },
        pokemon: {
          items: [],
          loading: false,
          refreshing: false,
          error: 'Gagal memuat data',
          nextUrl: null,
        },
      }),
    );
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<HomeScreen />);
    });
    const json = tr!.toJSON();
    const hasRetry = (node: any): boolean => {
      if (!node) return false;
      if (typeof node === 'string') return node === 'Coba Lagi';
      const children = node.children || [];
      return Array.isArray(children) && children.some(hasRetry);
    };
    expect(hasRetry(json)).toBe(true);
    act(() => tr!.unmount());
  });
});
