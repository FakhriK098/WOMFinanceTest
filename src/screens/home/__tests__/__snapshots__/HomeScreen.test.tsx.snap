// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders error state with retry (snapshot) 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#FFFFFF",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "padding": 20,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#6B7280",
              "fontSize": 14,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        Gagal memuat data
      </Text>
      <View
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
      />
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#4F46E5",
              "borderRadius": 12,
              "paddingVertical": 14,
            },
            undefined,
            undefined,
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 16,
                  "fontWeight": "700",
                },
                undefined,
              ]
            }
          >
            Coba Lagi
          </Text>
        </View>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 16,
            "elevation": 4,
            "height": 56,
            "justifyContent": "center",
            "padding": 8,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.2,
            "shadowRadius": 6,
            "width": 56,
          },
          {
            "backgroundColor": "#4F46E5",
          },
          {
            "bottom": 20,
            "position": "absolute",
            "right": 20,
          },
        ]
      }
    >
      <Image
        resizeMode="contain"
        source={
          {
            "testUri": "../../../src/assets/logout.png",
          }
        }
        style={
          [
            {
              "height": 24,
              "width": 24,
            },
            {
              "tintColor": "#FFFFFF",
            },
          ]
        }
      />
    </View>
  </View>
</RNCSafeAreaView>
`;

exports[`HomeScreen shows initial loading state (snapshot) 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  >
    <ActivityIndicator
      size="large"
    />
  </View>
</RNCSafeAreaView>
`;
